alf/utils/value_ops.py

```diff
@@ -144,39 +144,34 @@ def discounted_return(rewards, values, step_types, discounts, time_major=True):
         values = values.transpose(0, 1)
         step_types = step_types.transpose(0, 1)
 
-    assert values.shape[0] >= 2, "The sequence length needs to be \
-                                at least 2. Got {s}".format(s=values.shape[0])
+    assert values.shape[0] >= 2, ("The sequence length needs to be "
+                                  "at least 2. Got {s}".format(
+                                      s=values.shape[0]))
 
-    discounts = discounts[1:]
-    rewards = rewards[1:]
-    final_value = values[-1]
-    values = values[:-1]
-
-    step_types = step_types[:-1]
     is_lasts = (step_types == StepType.LAST).to(dtype=torch.float32)
 
-    def discounted_return_fn(acc_discounted_reward, reward, value, is_last,
-                             discount):
-        acc_discounted_value = acc_discounted_reward * discount + reward
-        return is_last * value + (1 - is_last) * acc_discounted_value
+    rets = torch.zeros(rewards.shape, dtype=rewards.dtype)
+    rets[-1] = values[-1]
+
+    with torch.no_grad():
```

Comment:
Can remove `with torch.no_grad()` because you will detach anyway

Reply:
```with torch.no_grad()``` will save memory